hydra-node/src/Hydra/API/ServerOutput.hs

```diff
@@ -115,3 +121,56 @@ instance
     Greetings me -> Greetings <$> shrink me
     PostTxOnChainFailed p e -> PostTxOnChainFailed <$> shrink p <*> shrink e
     RolledBack -> []
+
+-- | Possible transaction formats in the api server output
+data TxDisplay
+  = TxCBOR
+  | TxJSON
+  deriving (Eq, Show)
+
+-- | Replaces the json encoded tx field with it's cbor representation.
+-- NOTE: we deliberately pattern match on all 'ServerOutput' constructors
+-- so that we don't forget to update this function if they change.
+prepareServerOutput ::
+  IsChainState tx =>
+  -- | Decide on tx representation
+  TxDisplay ->
+  -- | Server output
+  TimedServerOutput tx ->
+  -- | Final output
+  LBS.ByteString
+prepareServerOutput TxJSON response = encode response
+prepareServerOutput TxCBOR response =
+  case output response of
+    PeerConnected{} -> encodedResponse
+    PeerDisconnected{} -> encodedResponse
+    HeadIsInitializing{} -> encodedResponse
+    Committed{} -> encodedResponse
+    HeadIsOpen{} -> encodedResponse
+    HeadIsClosed{} -> encodedResponse
+    HeadIsContested{} -> encodedResponse
+    ReadyToFanout{} -> encodedResponse
+    HeadIsAborted{} -> encodedResponse
+    HeadIsFinalized{} -> encodedResponse
+    CommandFailed{clientInput} ->
+      case clientInput of
+        Init -> encodedResponse
+        Abort -> encodedResponse
+        Commit{} -> encodedResponse
+        NewTx{Hydra.API.ClientInput.transaction = tx} -> replacedResponse tx
+        GetUTxO -> encodedResponse
+        Close -> encodedResponse
+        Contest -> encodedResponse
+        Fanout -> encodedResponse
+    TxValid{Hydra.API.ServerOutput.transaction = tx} -> replacedResponse tx
+    TxInvalid{Hydra.API.ServerOutput.transaction = tx} -> replacedResponse tx
+    SnapshotConfirmed{} -> encodedResponse
+    GetUTxOResponse{} -> encodedResponse
+    InvalidInput{} -> encodedResponse
+    Greetings{} -> encodedResponse
+    PostTxOnChainFailed{} -> encodedResponse
+    RolledBack -> encodedResponse
+ where
+  encodedResponse = encode response
+  replacedResponse tx =
+    encodedResponse & atKey "transaction" ?~ String (pack . unpack $ toStrictByteString $ toCBOR tx)
```

Comment:
Smell: we're _hacking_ an existing data structure to replace, on flight, one thing for another just before sending it. Can't think of a better idea from the top of my head, though :(

Reply:
Yes this solution is what we call _monkey-patch_. It is quick and dirty one since it is not trivial to pick the correct json instance for the transaction at the call place :(